### PR TITLE
Adds Contour Operator Enhancement Proposal

### PR DESCRIPTION
**Pull request by danehans**
Adds an enhancement proposal for an operator to manage Contour.

/assign @knobunc @Miciah 
/cc @frobware @sgreene570 @mcurry-rh @knrc 

Supports [NE-363](https://issues.redhat.com/browse/NE-363)

### CODE REVIEW

File: enhancements/ingress/contour-operator.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+* Create user and developer documentation.
+* Create tooling to simplify building, running, testing, etc. the operator.
+* Create tests that reduce bugs in the code and allow for continuous integration.
+* Integrate the operator with the OpenShift toolchain, i.e. openshift/release, required
+for productization.
+* Provide manifests for deploying the operator.
+
+The following functionality is expected as part of the operator:
+
+* Introduce a CRD used by the operator to manage Contour and its dependencies.
```
[the remark this answers is not in the document]
Comment (danehans) [reply enhancements/ingress/contour-operator.md]: @derekwaynecarr thanks for reviewing the enhancement. Although the upstream operator will support the projectcontour.io/* APIs (e.g. HTTPProxy), OpenShift will not. Contour is being added to OpenShift to support Service APIs which will supersede the Contour API's. Service APIs and the Contour Operator API(s) will follow the [Kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/). Let me update the enhancement to reflect these details.

**Comment by danehans** [comment]
@derekwaynecarr commit `e67178f` updates the EP based on your feedback. Thanks again for reviewing.

**Comment by bbrowning** [comment]
@markusthoemmes @nak3 If Contour gets added to OpenShift to support the Kubernetes Service APIs, is the surface area of the Kubernetes Service APIs enough for all our Serverless / Knative needs? Or would we expect Knative integration with Contour to also specifically required Contour APIs?

I'd like to understand if this proposal gets us on a path to not needing to maintain our own ingress (Kourier) for Knative or if we believe something additional would be needed for our ingress/routing layer that this proposal would not satisfy.

**Comment by markusthoemmes** [comment]
@bbrowning Knative's net-contour makes heavy use of the HTTPProxy API (see https://github.com/knative-sandbox/net-contour/blob/master/pkg/reconciler/contour/resources/httpproxy.go) which seems to be explicitly excluded in this proposal currently. That wouldn't suffice to meet our needs on the Serverless side to drop Kourier.

**Comment by markusthoemmes** [comment]
@bbrowning maybe I misread this. AFAIK, Knative is watching Ingress v2 and trying to make sure that indeed it will be sufficient to handle Knative's networking needs. At that point we should be fine assuming that Contour implements Ingress v2.

**Comment by bbrowning** [comment]
@markusthoemmes Thanks - that's more my concern, yes. So it sounds like from the Serverless team's point-of-view, this enhancement request would be a net positive, giving us a path towards moving to a more common routing layer and moving away from maintaining our own routing layer just for Knative workloads?

**Comment by markusthoemmes** [comment]
Assuming that Ingress V2 indeed fulfills the needs of Knative, absolutely yes. While on the topic, is there any timeline on when all of this might fall into place? Are we talking weeks/months or years?

**Comment by Miciah** [comment]
Let's merge this PR and iterate on it.
/lgtm

File: enhancements/ingress/contour-operator.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+## Design Details
+
+TBD
+
+### Open Questions [optional]
+
+1. Should the operator be created upstream to allow for community-based development and support? An
+[issue](https://github.com/projectcontour/contour/issues/2187) has been created to address this question
+with the Contour developer community.
```
Comment (smarterclayton) [inline enhancements/ingress/contour-operator.md:168]: Open question I'd like added - answering "how would an openshift customer select an ingress controller of type contour" as a follow up enhancement would gate this leaving tech preview for me (i think)

**Pull request merged**